include/swift/RemoteInspection/DescriptorFinder.h

```diff
@@ -0,0 +1,54 @@
+//===--------------- DescriptorFinder.h -------------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_REFLECTION_DESCRIPTOR_FINDER_H
+#define SWIFT_REFLECTION_DESCRIPTOR_FINDER_H
+
+#include "llvm/ADT/StringRef.h"
+
+namespace swift {
+namespace reflection {
+
+class TypeRef;
+
+/// An abstract interface for a builtin type descriptor.
+struct BuiltinTypeDescriptorInterface {
+  const uint32_t Size;
+  const uint32_t Alignment;
+  const uint32_t Stride;
+  const uint32_t NumExtraInhabitants;
+  const bool IsBitwiseTakable;
+
+  BuiltinTypeDescriptorInterface(uint32_t Size, uint32_t Alignment,
+                                 uint32_t Stride, uint32_t NumExtraInhabitants,
+                                 bool IsBitwiseTakable)
+      : Size(Size), Alignment(Alignment), Stride(Stride),
+        NumExtraInhabitants(NumExtraInhabitants),
+        IsBitwiseTakable(IsBitwiseTakable) {}
+
+  virtual ~BuiltinTypeDescriptorInterface() {};
+
+  virtual llvm::StringRef getMangledTypeName() = 0;
+};
+
+/// Interface for finding type descriptors. Implementors may provide descriptors
+/// that live inside or outside reflection metadata.
+struct DescriptorFinder {
+  virtual ~DescriptorFinder(){};
+
+  virtual std::unique_ptr<BuiltinTypeDescriptorInterface>
```

Comment:
Performance-wise, you could design the base class to already have all the data members needed by all implementations, maybe with an opaque element for implementation-specific data, then you could pass a stack-allocated object through this interface and don't need to use unique_ptr here.

Reply:
Personally I don't think the performance benefits would be worth the extra complexity. There's already indirection going on today when using `RemoteRef`, and there will be a pretty big difference in size between the DWARF version and the reflection version, so accommodating both on the base class would look a bit ugly imo.

Reply:
If this is only used in the debugger and similar clients, I think that's reasonable. If it's used in-process then I'm not sure.

Reply:
This is only used for out of process inspection.